# Promise
<br/>

### 이 글의 목적
    - AJAX/2024-01-17 02_promise.md에서 Promise Chaining에 대하여 학습하였다.
    - AJAX/2024-01-18 02_async_await.md에서 async-await function에 대하여 학습하였다.
    - 이번 글을 통해 Promise Chaining에서 async-await function으로 변환하여 모듈화하는 방법에 대해 알아보고자 한다.
<br/>

### 1. Promise Chaining - resolve
- [코드 1-1]은 Promise Chaining을 사용한 코드에서 비동기 작업이 성공적으로 이루어졌을 경우의 코드이다.
- setTimeout()의 콜백 함수에서 resolve()를 호출한다.
#### [결과 1]은 [코드 1-1], [코드 1-2], [코드 1-3]에 대한 출력 결과이다.
#### [코드 1-1]
```javascript
function job1() {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('job1 ok!');
        }, 2000);
    });
}
```

#### [코드 1-2]
```javascript
function job2() {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('job2 ok!');
        }, 2000);
    });
}
```
#### [코드 1-3]
```javascript
job1()
.then((response) => {
    console.log('response1 /', response);
    return job2();
})
.catch((error) => {
    console.log('error1 / ', error);
    return Promise.reject();
})
.then((response) => {
    console.log('response2 /', response);
});
```
#### [결과 1]
```plaintext
response1 / job1 ok!
response2 / job2 ok!
```
#### 구현 원리
```plaintext
    1. job1()이 return하는 Promise를 통해 .then()메서드를 실행하고 job2()를 return한다.
    2. .then() 메서드가 return하는 job2()도 Promise를 return하기 때문에
       제일 아래에 있는 .then() 메서드로 chaining 되어 response를 받고 console에 출력한다.
```
<br/>

### 2. Promise Chaining - reject
- [코드 2]는 Promise Chaining을 사용한 코드에서 비동기 작업이 실패하였을 경우의 코드이다.
- setTimeout()의 콜백 함수에서 reject()를 호출한다.
#### [결과 2]는 [코드 2], [코드 1-2], [코드 1-3]에 대한 출력 결과이다.
#### [코드 2]
```javascript
function job1() {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            reject('job1 failed!');
        }, 2000);
    });
}
```
#### [결과 2]
```plaintext
error1 /  job1 failed!
Uncaught (in promise) 
```
#### 구현 원리
```plaintext
    1. job1()이 return하는 Promise를 통해 .catch()메서드를 실행하고 error를 console에 출력한다.
    2. .catch() 메서드도 Promise를 return하기 때문에
       제일 아래에 있는 .then() 메서드로 chaining 되지 않도록 reject('job1 failed!');를 호출한다.
```
<br/>

### 3. async-await
- [코드 3-1] - [코드 3-4]는 Promise Chaining에서 async-await으로 바꾼 코드이다.
#### [코드 3-1]
```javascript
async function asyncJob1() {
    return await new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('async job1 ok!');
        }, 2000);
    });
}
```
#### [코드 3-2]
```javascript
async function asyncJob2() {
    return await new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('async job2 ok!');
        }, 2000);
    });
}
```
#### [코드 3-3]
```javascript
async function mainFunc() {
    const response1 = await asyncJob1();
    console.log('response1 /', response1);
    const response2 = await asyncJob2();
    console.log('response2 /', response2);
}
```
#### [코드 3-4]
```javascript
mainFunc();
```
#### [결과 3]
```plaintext
response1 / async job1 ok!
response2 / async job2 ok!
```
#### --> asyncJob1(), asyncJob2()처럼 비동기 작업에 대한 function을 미리 만들어 놓고 필요할 때마다 호출하면 되니
#### Promise Chaining뿐만 아니라 async-await function도 잘 활용할 수 있어야 할 것이다.
<br/>
